// @ohos/hvigor-ohos-plugin/src/enum/merge-type-rule.js
"use strict";
/*
 * Copyright (c) Huawei Technologies Co., Ltd. 2021-2022. All rights reserved.
 *
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.IsNeedMergeRule = exports.UniqueKeyEqualsName = exports.IsMapFields = exports.HasPlaceholderOptions = exports.NeedDisposePlaceholder = exports.KeepOneOfThem = exports.KeepAbilityMergeList = exports.KeepModuleAttr = exports.KeepHapOptions = void 0;
/**
 * 只保留hap中配置的options
 */
var KeepHapOptions;
(function (KeepHapOptions) {
    KeepHapOptions["APP_OPTIONS"] = "app";
    KeepHapOptions["ARK_OPTIONS"] = "ark";
    KeepHapOptions["NETWORK_OPTIONS"] = "network";
    KeepHapOptions["DISTRO_OPTIONS"] = "distro";
})(KeepHapOptions = exports.KeepHapOptions || (exports.KeepHapOptions = {}));
/**
 * 只保留hap中配置的module中的属性
 */
var KeepModuleAttr;
(function (KeepModuleAttr) {
    KeepModuleAttr["PACKAGE"] = "package";
    KeepModuleAttr["NAME"] = "name";
    KeepModuleAttr["DESCRIPTION"] = "description";
    KeepModuleAttr["DEVICE_TYPE"] = "deviceType";
    KeepModuleAttr["SUPPORTED_MODES"] = "supportModes";
    KeepModuleAttr["SHORTCUTS"] = "shortcuts";
    KeepModuleAttr["DISTRO"] = "distro";
})(KeepModuleAttr = exports.KeepModuleAttr || (exports.KeepModuleAttr = {}));
/**
 * ability中需要合并取合集的List属性
 */
var KeepAbilityMergeList;
(function (KeepAbilityMergeList) {
    KeepAbilityMergeList["PERMISSION"] = "permissions";
    KeepAbilityMergeList["BACKGROUND_MODES"] = "backgroundModes";
    KeepAbilityMergeList["CONFIG_CHANGES"] = "configChanges";
    KeepAbilityMergeList["META_DATA"] = "metaData";
})(KeepAbilityMergeList = exports.KeepAbilityMergeList || (exports.KeepAbilityMergeList = {}));
/**
 * 当hap和har同时有该属性时，使用hap的属性
 * 当hap和har只有一个有该属性时，使用非空的属性
 */
var KeepOneOfThem;
(function (KeepOneOfThem) {
    KeepOneOfThem["ALLOW_COMPONENTS_PROXY"] = "allowComponentsProxy";
    KeepOneOfThem["ALLOW_CLASS_MAP"] = "allowClassMap";
})(KeepOneOfThem = exports.KeepOneOfThem || (exports.KeepOneOfThem = {}));
/**
 * 需要处理${bundleName}占位符的标签
 */
var NeedDisposePlaceholder;
(function (NeedDisposePlaceholder) {
    NeedDisposePlaceholder["PERMISSIONS"] = "permissions";
    NeedDisposePlaceholder["READ_PERMISSIONS"] = "readPermission";
    NeedDisposePlaceholder["WRITE_PERMISSIONS"] = "writePermission";
    NeedDisposePlaceholder["NAME"] = "name";
    NeedDisposePlaceholder["URI"] = "uri";
})(NeedDisposePlaceholder = exports.NeedDisposePlaceholder || (exports.NeedDisposePlaceholder = {}));
/**
 * 需要处理${bundleName}占位符的标签
 */
var HasPlaceholderOptions;
(function (HasPlaceholderOptions) {
    HasPlaceholderOptions["ABILITIES"] = "abilities";
    HasPlaceholderOptions["DEF_PERMISSIONS"] = "defPermissions";
    HasPlaceholderOptions["DEF_PERMISSION_GROUPS"] = "defPermissionGroups";
})(HasPlaceholderOptions = exports.HasPlaceholderOptions || (exports.HasPlaceholderOptions = {}));
/**
 * 遵循map类合并规则的标签
 */
var IsMapFields;
(function (IsMapFields) {
    IsMapFields["MERGE_RULE"] = "mergeRule";
    IsMapFields["DEVICE_CONFIG"] = "deviceConfig";
})(IsMapFields = exports.IsMapFields || (exports.IsMapFields = {}));
/**
 * uniqueKey为name的options list
 */
var UniqueKeyEqualsName;
(function (UniqueKeyEqualsName) {
    UniqueKeyEqualsName["DOMAINS"] = "domains";
    UniqueKeyEqualsName["PARAMETERS"] = "parameters";
    UniqueKeyEqualsName["RESULTS"] = "results";
    UniqueKeyEqualsName["CUSTOMIZE_DATA"] = "customizeData";
    UniqueKeyEqualsName["COMMON_EVENTS"] = "commonEvents";
    UniqueKeyEqualsName["JS"] = "js";
    UniqueKeyEqualsName["REQ_PERMISSIONS"] = "reqPermissions";
    UniqueKeyEqualsName["DEF_PERMISSIONS"] = "defPermissions";
    UniqueKeyEqualsName["DEF_PERMISSION_GROUPS"] = "defPermissionGroups";
})(UniqueKeyEqualsName = exports.UniqueKeyEqualsName || (exports.UniqueKeyEqualsName = {}));
/**
 * 需要mergeRule的标签
 */
var IsNeedMergeRule;
(function (IsNeedMergeRule) {
    IsNeedMergeRule["ABILITIES"] = "abilities";
    IsNeedMergeRule["COMMON_EVENTS"] = "commonEvents";
    IsNeedMergeRule["DEF_PERMISSIONS"] = "defPermissions";
    IsNeedMergeRule["DEF_PERMISSION_GROUPS"] = "defPermissionGroups";
    IsNeedMergeRule["JS"] = "js";
    IsNeedMergeRule["REQ_PERMISSIONS"] = "reqPermissions";
    IsNeedMergeRule["CUSTOMIZE_DATA"] = "customizeData";
    IsNeedMergeRule["PARAMETERS"] = "parameters";
    IsNeedMergeRule["RESULTS"] = "results";
})(IsNeedMergeRule = exports.IsNeedMergeRule || (exports.IsNeedMergeRule = {}));
