// @ohos/hvigor-ohos-plugin/src/builder/inner-java-command-builder/d8-options.js
"use strict";
/*
 * Copyright (c) Huawei Technologies Co., Ltd. 2021-2022. All rights reserved.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.D8Options = void 0;
const java_command_builder_js_1 = require("../java-command-builder.js");
class D8Options extends java_command_builder_js_1.JavaCommandBuilder {
    constructor() {
        super();
    }
    addInputFiles(inputFile) {
        this.commandList.push(inputFile);
        return this;
    }
    addInputDir(inputDir) {
        this.commandList.push(`@${inputDir}`);
        return this;
    }
    addOutputDir(outputDir) {
        this.commandList.push('--output');
        this.commandList.push(outputDir);
        return this;
    }
    addLib(lib) {
        this.commandList.push('--lib');
        this.commandList.push(lib);
        return this;
    }
    addMinApi(api) {
        this.commandList.push('--min-api');
        this.commandList.push(api);
        return this;
    }
}
exports.D8Options = D8Options;
