// @ohos/hvigor-ohos-plugin/src/builder/inner-java-command-builder/packing-tool-options.js
"use strict";
/*
 * Copyright (c) Huawei Technologies Co., Ltd. 2021-2022. All rights reserved.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.PackingToolOptions = void 0;
const java_command_builder_js_1 = require("../java-command-builder.js");
class PackingToolOptions extends java_command_builder_js_1.JavaCommandBuilder {
    addMode(mode) {
        this.commandList.push('--mode');
        this.commandList.push(mode);
        return this;
    }
    addHapPath(hapPath) {
        this.commandList.push('--hap-path');
        this.commandList.push(hapPath);
        return this;
    }
    addHarPath(harPath) {
        this.commandList.push('--har-path');
        this.commandList.push(harPath);
        return this;
    }
    addPackInfoPath(packInfoPath) {
        this.commandList.push('--pack-info-path');
        this.commandList.push(packInfoPath);
        return this;
    }
    addOutPath(outPath) {
        this.commandList.push('--out-path');
        this.commandList.push(outPath);
        return this;
    }
    addFilePath(filePath) {
        this.commandList.push('--file-path');
        this.commandList.push(filePath);
        return this;
    }
    addBinPath(binPath) {
        this.commandList.push('--bin-path');
        this.commandList.push(binPath);
        return this;
    }
    addJsonPath(jsonPath) {
        this.commandList.push('--json-path');
        this.commandList.push(jsonPath);
        return this;
    }
    addPackResPath(packResPath) {
        if (packResPath === undefined) {
            return this;
        }
        this.commandList.push('--pack-res-path');
        this.commandList.push(packResPath);
        return this;
    }
    addEntryCardPath(entryCardFile) {
        if (entryCardFile === undefined) {
            return this;
        }
        this.commandList.push('--entrycard-path');
        this.commandList.push(entryCardFile);
        return this;
    }
    addResourcesPath(resourcesPath) {
        this.commandList.push('--resources-path');
        this.commandList.push(resourcesPath);
        return this;
    }
    addResourceTableTxtPath(txtPath) {
        this.commandList.push('--txt-path');
        this.commandList.push(txtPath);
        return this;
    }
    addJsPath(jsPath) {
        this.commandList.push('--js-path');
        this.commandList.push(jsPath);
        return this;
    }
    addEtsPath(etsPath) {
        this.commandList.push('--ets-path');
        this.commandList.push(etsPath);
        return this;
    }
    addAssetsPath(assetPath) {
        this.commandList.push('--assets-path');
        this.commandList.push(assetPath);
        return this;
    }
    addIndexPath(indexPath) {
        this.commandList.push('--index-path');
        this.commandList.push(indexPath);
        return this;
    }
    addLibPath(libPath) {
        this.commandList.push('--lib-path');
        this.commandList.push(libPath);
        return this;
    }
    addJarPath(jarPath) {
        this.commandList.push('--jar-path');
        this.commandList.push(jarPath);
        return this;
    }
    addSysCapPath(sysCapPath) {
        this.commandList.push('--rpcid-path');
        this.commandList.push(sysCapPath);
        return this;
    }
    force(force) {
        this.commandList.push('--force');
        this.commandList.push(force.toString());
        return this;
    }
}
exports.PackingToolOptions = PackingToolOptions;
