// @ohos/hvigor-ohos-plugin/src/builder/inner-java-command-builder/proguard-options.js
"use strict";
/*
 * Copyright (c) Huawei Technologies Co., Ltd. 2021-2022. All rights reserved.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.ProguardOptions = void 0;
const java_command_builder_js_1 = require("../java-command-builder.js");
class ProguardOptions extends java_command_builder_js_1.JavaCommandBuilder {
    constructor() {
        super();
    }
    inJars(path) {
        this.commandList.push('-injars');
        this.commandList.push(path);
        return this;
    }
    outJars(path) {
        this.commandList.push('-outjars');
        this.commandList.push(path);
        return this;
    }
    libraryJars(path) {
        this.commandList.push('-libraryjars');
        this.commandList.push(path);
        return this;
    }
    filePath(path) {
        this.commandList.push(`@${path}`);
        return this;
    }
    dontnote(filter) {
        this.commandList.push('-dontnote');
        this.commandList.push(filter);
        return this;
    }
}
exports.ProguardOptions = ProguardOptions;
