// @ohos/hvigor-ohos-plugin/src/model/ability/legacy-ability-model-impl.js
"use strict";
/*
 * Copyright (c) Huawei Technologies Co., Ltd. 2021-2021. All rights reserved.
 *
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.LegacyAbilityModelImpl = void 0;
const project_file_reader_js_1 = require("../../utils/project-file-reader.js");
const ability_type_enum_js_1 = require("../../enum/ability-type-enum.js");
const ohos_logger_js_1 = require("../../utils/log/ohos-logger.js");
/**
 * Fa模型中的ability数据模型
 *
 * @since 2022/1/19
 */
class LegacyAbilityModelImpl {
    constructor(configJson, abilityName) {
        this._log = ohos_logger_js_1.OhosLogger.getLogger(LegacyAbilityModelImpl.name);
        this._configObj = project_file_reader_js_1.ProjectFileReader.getJson5Obj(configJson);
        this._configFile = configJson;
        this._abilityName = abilityName;
        this._abilityObj = this.findAbilityByName();
        this._jsObj = this.findJsByName();
    }
    findJsByName() {
        if (!this._abilityObj || !this.hasJsObj()) {
            return undefined;
        }
        const jsObjs = this._configObj.module.js;
        for (let i = 0; i < jsObjs.length; i++) {
            if (jsObjs[i].name !== this._abilityName) {
                continue;
            }
            return jsObjs[i];
        }
        this._log._buildError(`Missing js tag for ability: ${this._abilityName}`)
            ._file(this._configFile)
            ._printErrorAndExit();
        return undefined;
    }
    hasJsObj() {
        // page需要有对应js页面
        return this.getType() === ability_type_enum_js_1.AbilityTypeEnum.PAGE;
    }
    findAbilityByName() {
        const abilityObjs = this._configObj.module.abilities;
        for (let i = 0; i < abilityObjs.length; i++) {
            if (abilityObjs[i].name !== this._abilityName) {
                continue;
            }
            return abilityObjs[i];
        }
        throw new Error(`在config.json module-ability中找不到${this._abilityName}`);
    }
    getRelateSrcPath() {
        return this._abilityObj.srcPath;
    }
    getName() {
        return this._abilityName;
    }
    getSrcLanguage() {
        return this._abilityObj.srcLanguage;
    }
    getType() {
        return ability_type_enum_js_1.AbilityTypeEnum.valueOf(this._abilityObj);
    }
    getConfigJsonJsObj() {
        return this._jsObj;
    }
}
exports.LegacyAbilityModelImpl = LegacyAbilityModelImpl;
