// @ohos/hvigor/src/options/cli-options.js
"use strict";
/*
 * Copyright (c) Huawei Technologies Co., Ltd. 2021-2021. All rights reserved.
 *
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.cliOptions = void 0;
/**
 * 提供Gvigor-cli的命令行options
 *
 * @since 2021/12/09
 */
exports.cliOptions = {
    help: {
        alias: 'h',
        type: 'boolean',
        describe: 'Show this help.',
    },
    version: {
        alias: 'v',
        type: 'boolean',
        describe: 'Print the global and local vigor versions.',
    },
    cwd: {
        type: 'string',
        requiresArg: true,
        describe: 'Manually set the CWD. The search for the vigorfile, ' +
            'as well as the relativity of all requires will be from here.'
    },
    require: {
        type: 'string',
        requiresArg: true,
        describe: 'Will require a module before running the vigorfile. ' +
            'This is useful for transpilers but also has other applications.',
    },
    prop: {
        alias: 'p',
        type: 'string',
        describe: 'Define extra properties.'
    },
    mode: {
        alias: "m",
        type: "string",
        describe: "Specifies the mode in which the command is currently executed."
    },
    sync: {
        alias: "s",
        type: "boolean",
        describe: "Sync the information in plugin for other platform."
    },
    error: {
        alias: "e",
        type: 'boolean',
        describe: 'Log errors only.'
    },
    warn: {
        alias: "w",
        type: 'boolean',
        describe: 'Set log level to warn.'
    },
    info: {
        alias: "i",
        type: 'boolean',
        describe: 'Set log level to info.'
    },
    debug: {
        alias: "d",
        type: 'boolean',
        describe: 'Set log level to debug.'
    }
};
