// @ohos/hvigor-ohos-plugin/src/tasks/legacy-tasks/legacy-generate-har-js-manifest.js
"use strict";
/*
 * Copyright (c) Huawei Technologies Co., Ltd. 2021-2022. All rights reserved.
 *
 */
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.LegacyGenerateHarJsManifest = void 0;
const path = __importStar(require("path"));
const fse = __importStar(require("fs-extra"));
const ohos_hap_task_js_1 = require("../task/ohos-hap-task.js");
const build_directory_const_js_1 = require("../../const/build-directory-const.js");
const hvigor_base_1 = require("@ohos/hvigor-base");
const inject_const_js_1 = require("../../const/inject-const.js");
const ohos_logger_js_1 = require("../../utils/log/ohos-logger.js");
const PAGE = inject_const_js_1.InjectConst.PREVIEWER_REPLACE_PAGE;
/**
 * 为har生成一个假的manifest
 *
 * @since 2022/3/1
 */
class LegacyGenerateHarJsManifest extends ohos_hap_task_js_1.OhosHapTask {
    constructor(taskService) {
        super(taskService, "LegacyGenerateHarJsManifest");
        this._log = ohos_logger_js_1.OhosLogger.getLogger(LegacyGenerateHarJsManifest.name);
        this.previewPage = hvigor_base_1.vigorConfigInst.getExtraConfig().get(PAGE);
    }
    static getWindowObj() {
        return {
            autoDesignWidth: false,
            designWidth: 454
        };
    }
    doTaskAction(targetData, target) {
        const moduleModel = this.service.getModuleModel();
        this.generateJsonInfo(moduleModel.getJsonObjByTargetName(target), targetData.getPathInfo());
    }
    getManifestPages() {
        if (!this.previewPage) {
            return [];
        }
        this._log.debug(`generate ${this.previewPage} for manifest.`);
        return [this.previewPage];
    }
    generateJsonInfo(configJsonObj, targetPathInfo) {
        const parentProject = this.service.getModuleModel().getParentProject();
        const manifestObj = {
            appID: configJsonObj.app.bundleName,
            versionName: configJsonObj.app.version.name,
            versionCode: configJsonObj.app.version.code,
            minPlatformVersion: parentProject.getProfileOpt().app.compatibleSdkVersion,
            appName: "",
            deviceType: configJsonObj.module.deviceType,
            window: LegacyGenerateHarJsManifest.getWindowObj(),
            pages: this.getManifestPages()
        };
        this._log._printDebugCommand("JsManifest", manifestObj);
        const manifestFile = path.resolve(targetPathInfo.getIntermediatesLegacyManifestJson(), build_directory_const_js_1.BuildArtifactConst.LEGACY_MANIFEST_JSON);
        fse.outputJSONSync(manifestFile, manifestObj);
    }
}
exports.LegacyGenerateHarJsManifest = LegacyGenerateHarJsManifest;
