// @ohos/hvigor-ohos-plugin/src/tasks/legacy-tasks/legacy-generate-js-manifest.js
"use strict";
/*
 * Copyright (c) Huawei Technologies Co., Ltd. 2021-2022. All rights reserved.
 *
 */
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.LegacyGenerateJsManifest = void 0;
const path = __importStar(require("path"));
const fse = __importStar(require("fs-extra"));
const ohos_hap_task_js_1 = require("../task/ohos-hap-task.js");
const build_directory_const_js_1 = require("../../const/build-directory-const.js");
const hvigor_base_1 = require("@ohos/hvigor-base");
const inject_const_js_1 = require("../../const/inject-const.js");
const ohos_logger_js_1 = require("../../utils/log/ohos-logger.js");
const ability_type_enum_js_1 = require("../../enum/ability-type-enum.js");
const SRC_PATH = inject_const_js_1.InjectConst.PREVIEWER_REPLACE_SRC_PATH;
const PAGE = inject_const_js_1.InjectConst.PREVIEWER_REPLACE_PAGE;
/**
 * 遍历project的module信息并生成JsManifest.json文件
 *
 * @since 2022/1/19
 */
class LegacyGenerateJsManifest extends ohos_hap_task_js_1.OhosHapTask {
    constructor(taskService) {
        super(taskService, "GenerateJsManifest");
        this._log = ohos_logger_js_1.OhosLogger.getLogger(LegacyGenerateJsManifest.name);
        this.previewSrcPath = hvigor_base_1.vigorConfigInst.getExtraConfig().get(SRC_PATH);
        this.previewPage = hvigor_base_1.vigorConfigInst.getExtraConfig().get(PAGE);
    }
    doTaskAction(targetData, target) {
        const moduleModel = this.service.getModuleModel();
        const abilityObjs = moduleModel.getLegacyAbilities(target);
        for (let i = 0; i < abilityObjs.length; i++) {
            const abilityInfo = abilityObjs[i];
            if (abilityInfo.getType() !== ability_type_enum_js_1.AbilityTypeEnum.PAGE &&
                abilityInfo.getType() !== ability_type_enum_js_1.AbilityTypeEnum.FORM) {
                continue;
            }
            this.generateJsonInfo(abilityObjs[i], moduleModel.getJsonObjByTargetName(target), targetData.getPathInfo());
        }
    }
    getManifestPages(legacyAbilityInfo) {
        var _a;
        if (!this.previewSrcPath || !this.previewPage || this.previewSrcPath !== legacyAbilityInfo.getRelateSrcPath()) {
            const pages = (_a = legacyAbilityInfo.getConfigJsonJsObj()) === null || _a === void 0 ? void 0 : _a.pages;
            return pages ? pages : [];
        }
        this._log.debug(`replace ${this.previewPage} for ${this.previewSrcPath} manifest.`);
        return [this.previewPage];
    }
    generateJsonInfo(legacyAbilityInfo, configJsonObj, targetPathInfo) {
        var _a, _b;
        const parentProject = this.service.getModuleModel().getParentProject();
        let manifestObj = {
            appID: configJsonObj.app.bundleName,
            versionName: configJsonObj.app.version.name,
            versionCode: configJsonObj.app.version.code,
            minPlatformVersion: parentProject.getProfileOpt().app.compatibleSdkVersion,
            appName: LegacyGenerateJsManifest.findAppName(configJsonObj),
            deviceType: configJsonObj.module.deviceType,
            window: LegacyGenerateJsManifest.getWindowObj(legacyAbilityInfo),
            pages: this.getManifestPages(legacyAbilityInfo)
        };
        const mode = (_a = legacyAbilityInfo.getConfigJsonJsObj()) === null || _a === void 0 ? void 0 : _a.mode;
        const type = legacyAbilityInfo.getType() === ability_type_enum_js_1.AbilityTypeEnum.FORM ?
            ability_type_enum_js_1.AbilityTypeEnum.FORM : (_b = legacyAbilityInfo.getConfigJsonJsObj()) === null || _b === void 0 ? void 0 : _b.type;
        if (mode) {
            manifestObj = Object.assign(Object.assign({}, manifestObj), { mode: mode });
        }
        else if (type) {
            manifestObj = Object.assign(Object.assign({}, manifestObj), { type: type });
        }
        this._log._printDebugCommand("JsManifest", manifestObj);
        const manifestFile = path.resolve(targetPathInfo.getIntermediatesLegacyManifestJson(), legacyAbilityInfo.getRelateSrcPath(), build_directory_const_js_1.BuildArtifactConst.LEGACY_MANIFEST_JSON);
        fse.outputJSONSync(manifestFile, manifestObj);
    }
    static findAppName(configJsonObj) {
        var _a;
        // find main ability
        const abilities = configJsonObj.module.abilities;
        for (let i = 0; i < abilities.length; i++) {
            const ability = abilities[i];
            if (!(ability === null || ability === void 0 ? void 0 : ability.skills)) {
                continue;
            }
            if (((_a = ability === null || ability === void 0 ? void 0 : ability.skills[0]) === null || _a === void 0 ? void 0 : _a.actions[0]) === 'action.system.home') {
                return ability.name;
            }
        }
        if (abilities[0].label) {
            return abilities[0].label;
        }
        return abilities[0].name;
    }
    static getWindowObj(legacyAbilityInfo) {
        var _a;
        const window = (_a = legacyAbilityInfo.getConfigJsonJsObj()) === null || _a === void 0 ? void 0 : _a.window;
        return window ?
            window :
            {
                autoDesignWidth: false,
                designWidth: 454
            };
    }
}
exports.LegacyGenerateJsManifest = LegacyGenerateJsManifest;
