// @ohos/hvigor-ohos-plugin/src/builder/inner-java-command-builder/new-signing-options.js
"use strict";
/*
 * Copyright (c) Huawei Technologies Co., Ltd. 2021-2022. All rights reserved.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.NewSigningOptions = void 0;
const java_command_builder_js_1 = require("../java-command-builder.js");
/**
 * 新签名工具hap-sign-tool的签名options
 *
 * @since 2022/03/01
 */
class NewSigningOptions extends java_command_builder_js_1.JavaCommandBuilder {
    sign() {
        this.commandList.push('sign-app');
        return this;
    }
    addMode(mode) {
        this.commandList.push('-mode');
        this.commandList.push(mode);
        return this;
    }
    addProfileFile(profile) {
        this.commandList.push('-profileFile');
        this.commandList.push(profile);
        return this;
    }
    addSignAlg(signAlg) {
        this.commandList.push('-signAlg');
        this.commandList.push(signAlg);
        return this;
    }
    addKeyAlias(privateKey) {
        this.commandList.push('-keyAlias');
        this.commandList.push(privateKey);
        return this;
    }
    addKeyStoreFile(keyStoreFile) {
        this.commandList.push('-keystoreFile');
        this.commandList.push(keyStoreFile);
        return this;
    }
    addKeyStorePwd(keyStorePwd) {
        this.commandList.push('-keystorePwd');
        this.commandList.push(keyStorePwd);
        return this;
    }
    addKeyPwd(keyAliasPwd) {
        this.commandList.push('-keyPwd');
        this.commandList.push(keyAliasPwd);
        return this;
    }
    addAppCertFile(certPath) {
        this.commandList.push('-appCertFile');
        this.commandList.push(certPath);
        return this;
    }
    addInputFile(inputFile) {
        this.commandList.push('-inFile');
        this.commandList.push(inputFile);
        return this;
    }
    addOutputFile(outputFile) {
        this.commandList.push('-outFile');
        this.commandList.push(outputFile);
        return this;
    }
}
exports.NewSigningOptions = NewSigningOptions;
