// @ohos/hvigor-ohos-plugin/src/version/sdk-version.js
"use strict";
/*
 * Copyright (c) Huawei Technologies Co., Ltd. 2022-2022. All rights reserved.
 *
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.SdkVersion = void 0;
const ohos_logger_js_1 = require("../utils/log/ohos-logger.js");
class SdkVersion {
    constructor(version) {
        this._log = ohos_logger_js_1.OhosLogger.getLogger(SdkVersion.name);
        const versionList = version.split('.');
        if (versionList.length !== 4) {
            this._log.errorMessageExit('Invalid sdk version.');
        }
        this._major = parseInt(versionList[0]);
        this._minor = parseInt(versionList[1]);
        this._revision = parseInt(versionList[2]);
        this._build = parseInt(versionList[3]);
    }
    getMajor() {
        return this._major;
    }
    getMinor() {
        return this._minor;
    }
    getRevision() {
        return this._revision;
    }
    getBuild() {
        return this._build;
    }
    /**
     * 判断当前版本号是否高于传入的版本
     *
     * @param {SdkVersion} sdkVersion
     * @returns {boolean}
     */
    isHigherThan(sdkVersion) {
        if (this._major > sdkVersion.getMajor()) {
            return true;
        }
        if (this._minor > sdkVersion.getMinor()) {
            return true;
        }
        if (this._revision > sdkVersion.getRevision()) {
            return true;
        }
        return this._build > sdkVersion.getBuild();
    }
}
exports.SdkVersion = SdkVersion;
