// @ohos/hvigor-ohos-plugin/index.js
"use strict";
/*
 * Copyright (c) Huawei Technologies Co., Ltd. 2021-2022. All rights reserved.
 *
 */
Object.defineProperty(exports, "__esModule", { value: true });
const plugin_factory_js_1 = require("./src/plugin/factory/plugin-factory.js");
/**
 * Stage模型的hap插件
 *
 * @param module hvigorModule
 */
const hapTasks = (module) => {
    const plugin = plugin_factory_js_1.PluginFactory.getHapPlugin(module);
    return {
        plugin,
        assembleHap: plugin.assembleHap,
        clean: plugin.clean,
        compileNative: plugin.compileNative,
        sync: plugin.sync,
        buildPreviewerResource: plugin.buildPreviewerRes
    };
};
/**
 * Fa模型的hap插件
 *
 * @param module hvigorModule
 */
const legacyHapTasks = (module) => {
    const plugin = plugin_factory_js_1.PluginFactory.getHapPlugin(module, true);
    return {
        plugin,
        assembleHap: plugin.assembleHap,
        clean: plugin.clean,
        compileNative: plugin.compileNative,
        buildPreviewerResource: plugin.buildPreviewerRes,
        sync: plugin.sync
    };
};
/**
 * Stage模型的app插件
 *
 * @param module hvigorProject
 */
const appTasks = (module) => {
    const plugin = plugin_factory_js_1.PluginFactory.getAppPlugin(module);
    return {
        plugin,
        assembleApp: plugin.assembleApp,
        clean: plugin.clean,
        sync: plugin.sync
    };
};
/**
 * Fa模型的app插件
 *
 * @param module hvigorProject
 */
const legacyAppTasks = (module) => {
    const plugin = plugin_factory_js_1.PluginFactory.getAppPlugin(module, true);
    return {
        plugin,
        assembleApp: plugin.assembleApp,
        clean: plugin.clean,
        sync: plugin.sync
    };
};
/**
 * Stage模型的Har插件
 *
 * @param module hvigorModule
 */
const harTasks = (module) => {
    const plugin = plugin_factory_js_1.PluginFactory.getHarPlugin(module);
    return {
        plugin,
        clean: plugin.clean,
        assembleHar: plugin.assembleHar,
        assembleSubHar: plugin.assembleSubHar,
        buildPreviewerResource: plugin.buildHarPreviewerRes
    };
};
/**
 * Fa模型的Har插件
 *
 * @param module hvigorModule
 */
const legacyHarTasks = (module) => {
    const plugin = plugin_factory_js_1.PluginFactory.getHarPlugin(module, true);
    return {
        plugin,
        clean: plugin.clean,
        assembleHar: plugin.assembleHar,
        assembleSubHar: plugin.assembleSubHar,
        buildPreviewerResource: plugin.buildHarPreviewerRes
    };
};
module.exports = {
    hapTasks,
    legacyHapTasks,
    appTasks,
    legacyAppTasks,
    harTasks,
    legacyHarTasks
};
