// @ohos/sdkmanager-common/build/src/com/huawei/deveco/sdkmanager/core/loader/ComponentDto.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.MetaDto = exports.ComponentDto = void 0;
const Utils_1 = require("../util/Utils");
class ComponentDto {
    constructor() {
        this.apiVersion = 0;
        this.displayName = '';
        this.path = '';
        this.releaseType = '';
        this.version = '';
    }
    compareTo(other) {
        const comparePath = this.path.localeCompare(other.getPath());
        if (comparePath !== 0) {
            return comparePath;
        }
        const compareApiVersion = this.apiVersion - other.getApiVersion();
        if (compareApiVersion !== 0) {
            return compareApiVersion;
        }
        return (0, Utils_1.compareVersion)(this.version, other.getVersion());
    }
    getApiVersion() {
        return this.apiVersion;
    }
    getDisplayName() {
        return this.displayName;
    }
    getLocation() {
        return this.location;
    }
    getMeta() {
        return this.meta;
    }
    getPath() {
        return this.path;
    }
    getReleaseType() {
        return this.releaseType;
    }
    getVersion() {
        return this.version;
    }
    hasApiVersion() {
        return this.apiVersion > 0;
    }
}
exports.ComponentDto = ComponentDto;
class MetaDto {
    constructor() {
        this.metaVersion = '';
    }
    getMetaVersion() {
        return this.metaVersion;
    }
}
exports.MetaDto = MetaDto;
