// @ohos/hvigor-ohos-plugin/src/builder/aapt2-command-builder.js
"use strict";
/*
 * Copyright (c) Huawei Technologies Co., Ltd. 2021-2022. All rights reserved.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.Aapt2CommandBuilder = void 0;
const base_command_builder_js_1 = require("./base-command-builder.js");
class Aapt2CommandBuilder extends base_command_builder_js_1.BaseCommandBuilder {
    constructor(executor) {
        super([]);
        this.commandList.push(executor);
    }
    compile() {
        this.commandList.push('compile');
        return this;
    }
    link() {
        this.commandList.push('link');
        return this;
    }
    showDetail() {
        this.commandList.push('-v');
        return this;
    }
    addDirectory(dir) {
        this.commandList.push('--dir');
        this.commandList.push(dir);
        return this;
    }
    addOutputPath(outputPath) {
        this.commandList.push('-o');
        this.commandList.push(outputPath);
        return this;
    }
    addApkPath(apkPath) {
        this.commandList.push('-I');
        this.commandList.push(apkPath);
        return this;
    }
    addRFilePath(rFilePath) {
        this.commandList.push('--java');
        this.commandList.push(rFilePath);
        return this;
    }
    addManifestPath(manifest) {
        this.commandList.push('--manifest');
        this.commandList.push(manifest);
        return this;
    }
    addPackageId(packageId) {
        this.commandList.push('--package-id');
        this.commandList.push(packageId);
        return this;
    }
    addIntermediates(intermediates) {
        this.commandList.push(intermediates);
        return this;
    }
}
exports.Aapt2CommandBuilder = Aapt2CommandBuilder;
