// @ohos/hvigor-ohos-plugin/src/builder/java-command-builder.js
"use strict";
/*
 * Copyright (c) Huawei Technologies Co., Ltd. 2021-2022. All rights reserved.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.JavaCommandBuilder = void 0;
const base_command_builder_js_1 = require("./base-command-builder.js");
class JavaCommandBuilder extends base_command_builder_js_1.BaseCommandBuilder {
    constructor(javaCommandBuilder) {
        super([]);
        javaCommandBuilder ? this.addAllParams(javaCommandBuilder.commandList) : this.commandList.push("java");
    }
    //  index
    addCalledJarFile(jarFile) {
        if (1 === this.commandList.length) {
            this.commandList.push('-jar');
            this.commandList.push(jarFile);
        }
        else {
            const java = this.commandList.slice(0, 1);
            const rest = this.commandList.slice(1);
            this.commandList = [...java, ...['-jar', jarFile], ...rest];
        }
        return this;
    }
    addJvmOption(jvmOption) {
        this.commandList.push(jvmOption);
        return this;
    }
    addClassPath(classpath) {
        this.commandList.push('-classpath');
        this.commandList.push(classpath);
        return this;
    }
    addMainClass(mainClass) {
        this.commandList.push(mainClass);
        return this;
    }
}
exports.JavaCommandBuilder = JavaCommandBuilder;
