// @ohos/hvigor-ohos-plugin/src/builder/restool-command-builder.js
"use strict";
/*
 * Copyright (c) Huawei Technologies Co., Ltd. 2021-2022. All rights reserved.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.RestoolCommandBuilder = void 0;
const base_command_builder_js_1 = require("./base-command-builder.js");
class RestoolCommandBuilder extends base_command_builder_js_1.BaseCommandBuilder {
    constructor(executor) {
        super([]);
        this.commandList.push(executor);
    }
    addInputVariantDir(inputVariantDir) {
        this.commandList.push('-b');
        this.commandList.push(inputVariantDir);
        return this;
    }
    addInputDir(inputDir) {
        this.commandList.push('-i');
        this.commandList.push(inputDir);
        return this;
    }
    addJsonFile(jsonFile) {
        this.commandList.push('-j');
        this.commandList.push(jsonFile);
        return this;
    }
    addModulePackName(packName) {
        this.commandList.push('-p');
        this.commandList.push(packName);
        return this;
    }
    addOutputDir(outputDir) {
        this.commandList.push('-o');
        this.commandList.push(outputDir);
        return this;
    }
    addResTable(resTable) {
        this.commandList.push('-r');
        this.commandList.push(resTable);
        return this;
    }
    forceDelete() {
        this.commandList.push('-f');
        return this;
    }
    incremental() {
        this.commandList.push('-c');
        return this;
    }
    compressImage() {
        this.commandList.push('-z');
        return this;
    }
    addOutputBak(outputBak) {
        this.commandList.push('-n');
        this.commandList.push(outputBak);
        return this;
    }
    addModules(modules) {
        this.commandList.push('-m');
        this.commandList.push(modules);
        return this;
    }
}
exports.RestoolCommandBuilder = RestoolCommandBuilder;
