// @ohos/hvigor-ohos-plugin/src/model/ability/legacy-form-model-impl.js
"use strict";
/*
 * Copyright (c) Huawei Technologies Co., Ltd. 2021-2021. All rights reserved.
 *
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.LegacyFormModelImpl = void 0;
const code_type_enum_js_1 = require("../../enum/code-type-enum.js");
const project_file_reader_js_1 = require("../../utils/project-file-reader.js");
const ability_type_enum_js_1 = require("../../enum/ability-type-enum.js");
/**
 * Fa模型中的form数据模型
 *
 * @since 2022/3/08
 */
class LegacyFormModelImpl {
    constructor(configJson, jsObj) {
        this._configObj = project_file_reader_js_1.ProjectFileReader.getJson5Obj(configJson);
        this._jsObj = jsObj;
        this._jsName = jsObj.name;
    }
    getRelateSrcPath() {
        return this.getName();
    }
    getName() {
        return this._jsName;
    }
    getSrcLanguage() {
        return code_type_enum_js_1.CodeType.JS;
    }
    getType() {
        // require page
        return ability_type_enum_js_1.AbilityTypeEnum.PAGE;
    }
    getConfigJsonJsObj() {
        return this._jsObj;
    }
}
exports.LegacyFormModelImpl = LegacyFormModelImpl;
