// @ohos/hvigor-ohos-plugin/src/model/module/core-module-model-impl.js
"use strict";
/*
 * Copyright (c) Huawei Technologies Co., Ltd. 2021-2022. All rights reserved.
 *
 */
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.CoreModuleModelImpl = void 0;
const common_const_js_1 = require("../../const/common-const.js");
const path_1 = __importDefault(require("path"));
const module_type_enum_js_1 = require("../../enum/module-type-enum.js");
const project_file_reader_js_1 = require("../../utils/project-file-reader.js");
const fs = __importStar(require("fs-extra"));
const validate_util_js_1 = require("../../utils/validate-util.js");
const ohos_logger_js_1 = require("../../utils/log/ohos-logger.js");
/**
 * hvigor 工程的module模块的数据管理对象
 *
 * @since 2022/2/23
 */
class CoreModuleModelImpl {
    constructor(modulePath, moduleName, parentProject) {
        this._log = ohos_logger_js_1.OhosLogger.getLogger(CoreModuleModelImpl.name);
        this.targetSourceSetMap = new Map();
        this.name = moduleName;
        this.modulePath = path_1.default.resolve(modulePath);
        this.parentProject = parentProject;
        this.buildProfilePath = path_1.default.resolve(this.modulePath, this.getBuildProfileName());
        // module build-profile.json5 schema check
        const moduleBuildProfilePath = path_1.default.resolve(this.modulePath, common_const_js_1.CommonConst.PROFILE_JSON5);
        this.moduleBuildProfileCheck(moduleBuildProfilePath);
        this._sourceRootDir = path_1.default.resolve(modulePath, "src");
    }
    getSourceRootByTargetName(targetName = common_const_js_1.DefaultTargetConst.DEFAULT_TARGET) {
        if (this.targetSourceSetMap.has(targetName)) {
            return this.targetSourceSetMap.get(targetName).getSourceSetRoot();
        }
        return this.targetSourceSetMap.get(common_const_js_1.DefaultTargetConst.DEFAULT_TARGET).getSourceSetRoot();
    }
    getBuildProfileName() {
        return common_const_js_1.CommonConst.BUILD_FILE_NAME;
    }
    getName() {
        return this.name;
    }
    getProjectDir() {
        return this.modulePath;
    }
    getParentProject() {
        return this.parentProject;
    }
    getPackageJsonPath() {
        return path_1.default.resolve(this.modulePath, common_const_js_1.CommonConst.PACKAGE_JSON);
    }
    getProfilePath() {
        return path_1.default.resolve(this.modulePath, common_const_js_1.CommonConst.PROFILE_JSON5);
    }
    getProfileOpt() {
        return project_file_reader_js_1.ProjectFileReader.getJson5Obj(this.getProfilePath());
    }
    isArkModule() {
        var _a;
        const arkEnable = (_a = this.getProfileOpt().buildOption) === null || _a === void 0 ? void 0 : _a.arkEnable;
        return arkEnable === undefined ? true : arkEnable;
    }
    isHapModule() {
        return this.getModuleType() !== module_type_enum_js_1.ModuleType.Har;
    }
    isHarModule() {
        return this.getModuleType() === module_type_enum_js_1.ModuleType.Har;
    }
    getCompileApiVersion() {
        return this.getParentProject().getProfileOpt().app.compileSdkVersion;
    }
    getCompatibleApiVersion() {
        return this.getParentProject().getProfileOpt().app.compatibleSdkVersion;
    }
    moduleBuildProfileCheck(moduleProfilePath) {
        const moduleSchema = require('../../../res/schemas/ohos-module-build-profile-schema.json');
        if (!fs.existsSync(moduleProfilePath)) {
            this._log._buildError(`Can not find build config file build-profile.json5.`)
                ._file(moduleProfilePath)
                ._printErrorAndExit(this.name);
        }
        validate_util_js_1.ValidateUtil.doSchemaCheck(this.name, moduleProfilePath, moduleSchema);
    }
}
exports.CoreModuleModelImpl = CoreModuleModelImpl;
