// @ohos/hvigor-ohos-plugin/src/plugin/hap-plugin.js
"use strict";
/*
 * Copyright (c) Huawei Technologies Co., Ltd. 2021-2022. All rights reserved.
 *
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.HapPlugin = void 0;
const build_previewer_res_js_1 = require("../tasks/previewer/build-previewer-res.js");
const assemble_hap_js_1 = require("../tasks/assemble/assemble-hap.js");
const sync_module_js_1 = require("../tasks/sync/sync-module.js");
const compile_native_js_1 = require("../tasks/compile-native.js");
const clean_js_1 = require("../tasks/common/clean.js");
const abstract_module_plugin_js_1 = require("./common/abstract-module-plugin.js");
/**
 * 对外暴露hap级别的接口和任务的plugin
 *
 * @since 2021/12/16
 */
class HapPlugin extends abstract_module_plugin_js_1.AbstractModulePlugin {
    constructor(module, isFaMode) {
        super(module, isFaMode);
    }
    initTasks() {
        this.assembleHap = new assemble_hap_js_1.AssembleHap(this._taskService, this.isFaMode);
        this.clean = new clean_js_1.Clean(this._taskService);
        this.compileNative = new compile_native_js_1.CompileNative(this._taskService);
        this.sync = new sync_module_js_1.SyncModule(this._module);
        this.buildPreviewerRes = new build_previewer_res_js_1.BuildPreviewerRes(this._taskService, this.isFaMode);
    }
}
exports.HapPlugin = HapPlugin;
