// @ohos/hvigor-ohos-plugin/src/plugin/har-plugin.js
"use strict";
/*
 * Copyright (c) Huawei Technologies Co., Ltd. 2021-2022. All rights reserved.
 *
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.HarPlugin = void 0;
const assemble_har_js_1 = require("../tasks/assemble/assemble-har.js");
const assemble_sub_har_js_1 = require("../tasks/assemble/assemble-sub-har.js");
const clean_js_1 = require("../tasks/common/clean.js");
const abstract_module_plugin_js_1 = require("./common/abstract-module-plugin.js");
const build_har_previewer_res_js_1 = require("../tasks/previewer/build-har-previewer-res.js");
/**
 * 对外暴露har级别的接口和任务的plugin
 *
 * @since 2021/12/16
 */
class HarPlugin extends abstract_module_plugin_js_1.AbstractModulePlugin {
    constructor(module, isFaMode) {
        super(module, isFaMode);
    }
    initTasks() {
        this.assembleHar = new assemble_har_js_1.AssembleHar(this._taskService, this.isFaMode);
        this.assembleSubHar = new assemble_sub_har_js_1.AssembleSubHar(this._taskService, this.isFaMode);
        this.clean = new clean_js_1.Clean(this._taskService);
        this.buildHarPreviewerRes = new build_har_previewer_res_js_1.BuildHarPreviewerRes(this._taskService, this.isFaMode);
    }
}
exports.HarPlugin = HarPlugin;
